fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,546 @@
+                raise MOVING_DENIED
+            if torch.is_tensor(args[0]):
+                raise MOVING_DENIED
+
+        return super().to(*args, **kwargs)
 
     def parameter_rrefs(self) -> List[rpc.RRef]:
-        rrefs_list_of_lists = [rpc.rpc_sync(l.owner(), _parameter_rrefs, args=(l,)) for l in self.rmodule]
-        return list(itertools.chain(*rrefs_list_of_lists))
+        remote_params = []
+        for p in self.partition_handlers:
+            remote_params.extend(p.rpc_sync().local_parameter_rrefs())
+        print("params rrefs: ", len(remote_params))
+        return remote_params
+
+    def forward(self, *inputs: Tensor) -> rpc.RRef:  # type: ignore
+        for i, input in enumerate(inputs):
+            microbatch.check(input)
+
+        if not self.partition_handlers:
+            # Empty sequential module is not illegal.
+            assert len(inputs) == 1
+            return rpc.RRef(inputs[0])
+
+        # Divide a mini-batch into micro-batches.
+        batches_list = [microbatch.scatter(input, self.chunks) for input in inputs]
+
+        m = len(self.partition_handlers)
+        dist_records: List[Optional[rpc.RRef]] = [None] * (m + 1)
+        for part_idx in reversed(range(m)):
+            rmodel = self.partition_handlers[part_idx].remote()
+            users = []
+            for user, input_idx, output_idx in self.graph.output_users[self.partitions[part_idx][0][-1]]:
+                user_partition = next(i for i, (p, m) in enumerate(self.partitions) if p[0] == user)
+                assert user_partition > part_idx
+                users.append((dist_records[user_partition], input_idx, output_idx))
+            dist_records[part_idx] = rmodel.make_dist_record(users)
+            this_result = rmodel.run_pipeline(dist_records[part_idx])
+            if part_idx == m - 1:
+                result = this_result
+
+        for i, b in enumerate(zip(*batches_list)):
+            for fi, fj, feed_idx in self.input_feeds:
+                # TODO: Debug why we need this special handling
+                dist_record = dist_records[fi]
+                assert dist_record is not None
+                if dist_record.owner().name == rpc.get_worker_info().name:  # type: ignore
+                    dist_record.local_value().feed(i, fj, b[feed_idx].value)
+                else:
+                    dist_record.rpc_async().feed(i, fj, b[feed_idx].value)  # type: ignore
+
+        return result
+
+
+def create_sequence_pipeline(
+    layers: List[PipelineModule], balance: List[int], devices: List[str], **kwargs: Any
+) -> DistributedPipeline:
+    g = PipelineModulesGraph()
+
+    index = 0
+    for num_layers, remote_device in zip(balance, devices):
+        next_index = index + num_layers
+        for li in range(index, next_index):
+            layers[li].instantiate(remote_device)
+        index = next_index
+
+    g.add_sequence(layers)
+    g.feed_model_input(layers[0])
```

Comment:
I think the PipelineModulesGraph code would be clearer if add_sequence and feed_model_input were called directly from the PipelineModulesGraph constructor.

Reply:
Not sure if I understand the design in your mind. PipelineModulesGraph is a utility to create arbitrary acyclic graph. It is very hard to inject all necessary information as constructor argument.